src/xmipp/libraries/data/monogenic.cpp

```diff
@@ -0,0 +1,712 @@
+	sdN2 = sumN2/NN - meanN*meanN;
+}
+
+
+// SETLOCALRESOLUTIONHALFMAPS: Set the local resolution of a voxel, by
+// determining if the monogenic amplitude "amplitudeMS" is higher than
+// the threshold of noise "thresholdNoise". Thus the local resolution
+// map "plocalResolution" is set, with the resolution value "resolution"
+// or with "resolution_2". "resolution" is the resolution for with
+// the hypohtesis test is carried out, and "resolution_2" is the resolution
+// of the analisys two loops ago (see monores method)
+void Monogenic::setLocalResolutionHalfMaps(const MultidimArray<double> &amplitudeMS,
+		MultidimArray<int> &pMask, MultidimArray<double> &plocalResolutionMap,
+		double &thresholdNoise, double &resolution, double &resolution_2)
+	{
+		FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(amplitudeMS)
+		{
+			if (DIRECT_MULTIDIM_ELEM(pMask, n)>=1)
+				if (DIRECT_MULTIDIM_ELEM(amplitudeMS, n)>thresholdNoise)
+				{
+					DIRECT_MULTIDIM_ELEM(pMask, n) = 1;
+					DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution;
+				}
+				else{
+					DIRECT_MULTIDIM_ELEM(pMask, n) += 1;
+					if (DIRECT_MULTIDIM_ELEM(pMask, n) >2)
+					{
+						DIRECT_MULTIDIM_ELEM(pMask, n) = 0;
+						DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution_2;
+					}
+				}
+		}
+	}
+
+// SETLOCALRESOLUTIONHALFMAPS: Set the local resolution of a voxel, by
+// determining if the monogenic amplitude "amplitudeMS" is higher than
+// the threshold of noise "thresholdNoise". Thus the local resolution
+// map "plocalResolution" is set, with the resolution value "resolution"
+// or with "resolution_2". "resolution" is the resolution for with
+// the hypohtesis test is carried out, and "resolution_2" is the resolution
+// of the analisys two loops ago (see monores method)
+void Monogenic::setLocalResolutionMap(const MultidimArray<double> &amplitudeMS,
+		MultidimArray<int> &pMask, MultidimArray<double> &plocalResolutionMap,
+		double &thresholdNoise, double &resolution, double &resolution_2)
+	{
+		FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(amplitudeMS)
+		{
+			if (DIRECT_MULTIDIM_ELEM(pMask, n)>=1)
+				if (DIRECT_MULTIDIM_ELEM(amplitudeMS, n)>thresholdNoise)
+				{
+					DIRECT_MULTIDIM_ELEM(pMask, n) = 1;
+					DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution;//sampling/freq;
+				}
+				else{
+					DIRECT_MULTIDIM_ELEM(pMask, n) += 1;
+					if (DIRECT_MULTIDIM_ELEM(pMask, n) >2)
+					{
+						DIRECT_MULTIDIM_ELEM(pMask, n) = -1;
+						DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution_2;//maxRes - counter*R_;
+					}
+				}
+		}
+	}
+
+
+// MONOGENICAMPLITUDE_3D_FOURIER: Given the fourier transform of a map
+// "myfftV", this function computes the monogenic amplitude "amplitude" 
+// iu is the inverse of the frequency in Fourier space.
+void Monogenic::monogenicAmplitude_3D_Fourier(const MultidimArray< std::complex<double> > &myfftV,
+		MultidimArray<double> iu, MultidimArray<double> &amplitude, int numberOfThreads)
```

Comment:
`iu` should be probably reference, or not passed at all, as you alter it locally

Reply:
Absolutely, my fault!, done